projects/rocr-runtime/runtime/hsa-runtime/image/hsa_amd_mipmap.cpp

```diff
@@ -0,0 +1,885 @@
+      base_dim = std::max<uint32_t>(desc->width, desc->height);
+  } else if (desc->geometry == HSA_EXT_IMAGE_GEOMETRY_3D) {
+      base_dim = std::max<uint32_t>(base_dim, std::max<uint32_t>(desc->height, desc->depth));
+  }
+  uint32_t max_levels = 1;
+  if (base_dim > 1) {
+      max_levels = 1 + static_cast<uint32_t>(std::floor(std::log2(static_cast<double>(base_dim))));
+  }
+  if (num_mipmap_levels > max_levels) {
+      return HSA_STATUS_ERROR_INVALID_ARGUMENT;
+  }
+
+  auto* rt = rocr::image::ImageRuntime::instance();
+  if (!rt) return HSA_STATUS_ERROR_OUT_OF_RESOURCES;
+
+  return rt->CreateMipmapArrayHandle(agent, *desc, image_data, image_data_size,
+                                   access_permission, num_mipmap_levels, layout,
+                                   row_pitch, slice_pitch, *image_out);
+
+  CATCH
+}
+
+hsa_status_t HSA_API
+hsa_amd_mipmap_array_destroy(const hsa_ext_image_t* image) {
+  TRY
+
+  if (!image) return HSA_STATUS_ERROR_INVALID_ARGUMENT;
+  if (image->handle == 0) return HSA_STATUS_ERROR_INVALID_ARGUMENT;
+
+  auto* rt = rocr::image::ImageRuntime::instance();
+  if (!rt) return HSA_STATUS_ERROR_OUT_OF_RESOURCES;
+
+  return rt->DestroyMipmapArrayHandle(*image);
+
+  CATCH
+}
+
+hsa_status_t HSA_API
+hsa_amd_mipmap_array_get_level(
+        hsa_agent_t agent,
+        const hsa_ext_image_t* mipmapped_array,
+        uint32_t mip_level,
+        hsa_ext_image_t* level_image_out) {
+    TRY
+
+    if (!mipmapped_array || !level_image_out) {
+        return HSA_STATUS_ERROR_INVALID_ARGUMENT;
+    }
+
+    if (mipmapped_array->handle == 0) {
+        return HSA_STATUS_ERROR_INVALID_ARGUMENT;
+    }
+
+    const rocr::image::MipmappedArray* array =
+        rocr::image::MipmappedArray::Convert(mipmapped_array->handle);
+    if (!array || mip_level >= array->num_levels) {
+        return HSA_STATUS_ERROR_INVALID_ARGUMENT;
+    }
+
+    // Create individual level image descriptor using AddressLib mip info
+    const ADDR2_MIP_INFO* mip_infos = array->addr_output.pMipInfo;
+    if (!mip_infos) {
+        return HSA_STATUS_ERROR_INVALID_ARGUMENT;
+    }
+
+    const ADDR2_MIP_INFO& lvl = mip_infos[mip_level];
+
+    // Create level descriptor with AddressLib computed dimensions
+    hsa_ext_image_descriptor_t level_desc = array->desc;
+    level_desc.width = lvl.pixelPitch ? lvl.pixelPitch :
+        std::max<uint32_t>(1u, array->desc.width >> mip_level);
+    level_desc.height = lvl.pixelHeight ? lvl.pixelHeight :
+        std::max<uint32_t>(1u, array->desc.height >> mip_level);
+
+    if (array->desc.geometry == HSA_EXT_IMAGE_GEOMETRY_3D) {
+        level_desc.depth = std::max<uint32_t>(1u, array->desc.depth >> mip_level);
+    }
+
+    // Calculate level data pointer using AddressLib offset
+    uint8_t* level_data = static_cast<uint8_t*>(array->data);
+    if (level_data && lvl.offset > 0) {
+        level_data += lvl.offset;
+        if (lvl.macroBlockOffset) level_data += lvl.macroBlockOffset;
+    }
```

Comment:
I really don't recommend creating an image sub-view in this manner; you *can* technically spoof an image of a given level directly but there are edge cases that get significantly more complicated than this. Instead, use the SRD 'base_level' and 'last_level' fields dedicated to this purpose.